src/process.rs

```diff
@@ -142,3 +147,96 @@ impl<T: Runtime> Future for ProcessController<T> {
         }
     }
 }
+
+pub struct LossyLinesCodec {
+    max_length: usize,
+}
+
+impl Default for LossyLinesCodec {
+    fn default() -> Self {
+        Self {
+            max_length: usize::MAX,
+        }
+    }
+}
+
+pub type OutputLines = Pin<Box<dyn futures::Stream<Item = anyhow::Result<String>> + Send>>;
+
+/// Reads process stdout and stderr using `LossyLinesCodec`
+pub fn process_output(child: &mut Child) -> anyhow::Result<OutputLines> {
+    let stdout = child
+        .stdout
+        .take()
+        .context("Failed to access process stdout")?;
+    let stderr = child
+        .stderr
+        .take()
+        .context("Failed to access process stderr")?;
+
+    let stdout = FramedRead::new(BufReader::new(stdout), LossyLinesCodec::default());
+    let stderr = FramedRead::new(BufReader::new(stderr), LossyLinesCodec::default());
+
+    Ok(futures::StreamExt::boxed(stdout.merge(stderr)))
+}
+
+/// Decodes lines as UTF-8 (lossly) up to `max_length` characters per line.
+impl Decoder for LossyLinesCodec {
+    type Item = String;
+
+    type Error = anyhow::Error;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        let read_to = std::cmp::min(self.max_length.saturating_add(1), buf.len());
+        let new_line_offset = buf[0..read_to].iter().position(|b| *b == b'\n');
+        let has_new_line = new_line_offset.is_some();
+        let offset = new_line_offset.unwrap_or(read_to);
+        let line = buf.split_to(offset);
+        if has_new_line {
+            // Move cursor pass new line character so next call of `decode` will not read it.
+            buf.advance(1);
+        }
+        let mut line: &[u8] = &line;
+        if let Some(&b'\r') = line.last() {
+            // Skip carriage return.
+            line = &line[..line.len() - 1];
+        }
+        if line.is_empty() {
+            return Ok(None);
+        }
+        // Lossy conversion to avoid errors when converting UTF-16 characters on Windows process output.
+        let line = String::from_utf8_lossy(line).to_string();
```

Comment:
I needed to implement my own codec, because original `LinesCodec` throws an Error on UTF-8 decode failure, at this stage bytebuffer is already advanced forward, and I found no way to fallback into some alternative decoding.

Reply:
I am a little confused - are we trying to read utf16 here with `from_utf8_lossy` and just assume the invalid characters will me marked as the �?

Reply:
In other words  shouldn't we rather read utf16 and convert to String?

Reply:
It happens on Windows. On Linux output is in UTF-8 so I would need to make this code OS aware. For that matter Windows can be UTF-16 BE and LE. I assumed it is not worth to handle it only to get string like this printed correctly:
```
[2024-05-23T12:49:32.443+0200 DEBUG ya_runtime_ai::process::automatic::monitor] .436+0200 DEBUG ya_runtime_ai::process::automatic::monitor] > OSError: [WinError 193] %1 nie jest prawid�ow� aplikacj� systemu Win32. Error loading "C:\Users\golem-dev\AppData\Local\GamerHashAI\modules\golem_4.1.0\plugins\ya-automatic-ai\automatic\system\python\lib\site-packages\torch\lib\c10.dll" or one of its dependencies.
```
Without this PR such message would be not included in logs at all.